cpp/src/arrow/python/numpy_to_arrow.cc

```diff
@@ -594,7 +594,13 @@ Status NumPyConverter::Visit(const FixedSizeBinaryType& type) {
 
   if (mask_ != nullptr) {
     Ndarray1DIndexer<uint8_t> mask_values(mask_);
-    RETURN_NOT_OK(builder.AppendValues(data, length_, mask_values.data()));
+    std::unique_ptr<uint8_t[]> inverted_mask(new uint8_t[length_]);
+    for (int64_t i = 0; i < length_; ++i) {
+      inverted_mask[i] = !mask_values[i];
+    }
```

Comment:
This is doing a spurious additional allocation and memory copy. You could reuse `MaskToBitmap` instead.

Reply:
The issue is that `AppendValues` needs a byte mask, not bit mask

Reply:
Right. That said, we can also eschew `AppendValues` and construct the `ArrayData` directly.

Reply:
The version for variable length binaries ( `NumPyConverter::Visit(const BinaryType& type)` ) does that same work simply iterating over the data and invoking `AppendNull` / `Append` depending on the value of the mask.  Would that approach be ok? That would avoid the need to allocate the inverted mask

Reply:
Yes, that would work. Note you can also call `builder.Reserve` to preallocate the required data size.

Reply:
I'm wondering, given that this seems a fairly reasonable need, if the ArrayBuilder itself shouldn't have a specialised version of `AppendValues` that accepts a mask. Maybe `AppendValuesWithMask`. 

It seems confusing by the way that the builder interface would have multiple different "definitions" of how to edit the bitmap, but in the end we already have such dichotomy in place as the ``Array`` **mask** is ``True for NULL`` while the **bitmap** is ``0 for NULL``

Reply:
Well, the only place where this is needed is the PyArrow API, and I don't even know why we have this `mask` argument (with the confusing convention that `true` means "null", which is exactly the reverse of Arrow validity bitmaps).

Reply:
The `mask` keyword is useful if you have numpy-style masked arrays (having a boolean numpy array as the mask). We use this keyword in pandas.

Reply:
With this change, we are now looping over the values, and appending iteratively. Would that have performance implications?

Since this is for a fixed size/strided array, an alternative could be to create the ArrayData directly (as mentioned by Antoine), which would be similar to VisitNative I suppose?

https://github.com/apache/arrow/blob/3a9aea3dd0e2a2b1c68bd9cfc141d83adb9fbc35/cpp/src/arrow/python/numpy_to_arrow.cc#L276-L280

Reply:
I didn't benchmark the code, but notice that we were iterating over a mask of the same length and this is the same thing we are also doing for strings and varlen binary

https://github.com/apache/arrow/blob/master/cpp/src/arrow/python/numpy_to_arrow.cc#L561-L567

I do see btw that ``ArrayData::Make`` is able to do a zero copy from numpy data under certain conditions, so it will surely be a lot faster if we end into that condition.  In the other case, it seems that we end into ``CopyStrided*`` which does iterate over the data like now and thus the performance benefit would probably be negligible if any.

Reply:
Ok, the implementation provided by https://github.com/apache/arrow/pull/10199/commits/b3e0a37049dffdcac63b5ce446907be60523e651 should do the trick.

I also verified that the `PrepareInputData` does a zero copy and `ArrayData::Make` does move the buffer.
Which is good because it should be far faster than previous solution, but raises questions about the memory ownership.

Reply:
Confirmed that this new implementation is fully sharing memory with numpy. 

```
        npa = np.array([b'aaa', b'bbb', b'ccc']*100)
        arrow_array = pa.array(npa, type=pa.binary(3),
                               mask=np.array([False, False, False]*100))
        npa[npa == b"bbb"] = b"XXX"
>       assert ([b'aaa', b'bbb', b'ccc']*100) == arrow_array.to_pylist()
E       AssertionError: assert [b'aaa', b'bb..., b'ccc', ...] == [b'aaa', b'XX..., b'ccc', ...]
E         At index 1 diff: b'bbb' != b'XXX'
```

Not sure that's what we want to do. It seems that when doing `pa.array(np.array)` it's currently pretty random if we end up sharing memory or not. 

For example 
```
npa = np.array([1, 2, 3]*3)
arrow_array = pa.array(npa, type=pa.int64())
```
shares memory, while
``` 
npa = np.array([1, 2, 3]*3)
arrow_array = pa.array(npa, type=pa.int32())
``` 
doesn't.

That should probably be consolidated.

Thinking about it, I don't think we should be sharing memory by default when building a new array, as the user expectation is probably that they are building something new.

Reply:
Created https://issues.apache.org/jira/browse/ARROW-12667 and https://issues.apache.org/jira/browse/ARROW-12666 as follow-up of this issue. Reverted the implementation to the previous one which was in line with the behaviour exposed for VariableLengthBinary and didn't have the memory copy issue.